pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java

```diff
@@ -745,58 +942,86 @@ public VmClass getVmClass() {
     }
 
     @Override
-    protected final PType doExport() {
+    protected PType doExport() {
       return new PType.Class(BaseModule.getCollectionClass().export(), elementTypeNode.doExport());
     }
 
     @Override
-    public final VmList getTypeArgumentMirrors() {
+    public VmList getTypeArgumentMirrors() {
       return VmList.of(elementTypeNode.getMirror());
     }
 
-    @Specialization
-    protected void eval(VirtualFrame frame, VmList value) {
+    @Override
+    protected void acceptTypeNode(Consumer<TypeNode> visitor) {
+      visitor.accept(this);
+    }
+
+    private Object evalList(VirtualFrame frame, VmList value) {
+      var ret = value;
+      var idx = 0;
+
       for (var elem : value) {
-        elementTypeNode.execute(frame, elem);
+        var result = elementTypeNode.execute(frame, elem);
+        if (result != elem) {
+          ret = ret.replace(idx, result);
+        }
+        idx++;
       }
 
-      LoopNode.reportLoopCount(this, value.getLength());
+      LoopNode.reportLoopCount(this, idx);
+      return ret;
     }
 
-    @Specialization
-    protected void eval(VirtualFrame frame, VmSet value) {
+    private Object evalListEagerly(VirtualFrame frame, VmList value) {
+      var ret = value;
+      var idx = 0;
+
       for (var elem : value) {
-        elementTypeNode.execute(frame, elem);
+        var result = elementTypeNode.executeEagerly(frame, elem);
+        if (result != elem) {
+          ret = ret.replace(idx, result);
+        }
+        idx++;
       }
 
       LoopNode.reportLoopCount(this, value.getLength());
+      return value;
     }
 
-    @Fallback
-    protected void fallback(Object value) {
-      throw typeMismatch(value, BaseModule.getCollectionClass());
+    private Object evalSet(VirtualFrame frame, VmSet value) {
+      for (var elem : value) {
+        elementTypeNode.executeEagerly(frame, elem);
+      }
+
+      LoopNode.reportLoopCount(this, value.getLength());
+      return value;
     }
   }
 
-  public abstract static class ListTypeNode extends ObjectSlotTypeNode {
+  public static final class ListTypeNode extends ObjectSlotTypeNode {
```

Comment:
Truffle's DSL was overriding `executeEagerly` when making this an abstract class and using specializations. I'm not sure if I'm doing this wrong?

But, regardless, having specializations here doesn't seem that useful anyway? Reading the same member value skips the typecheck the second time around because it gets cached. And reading a new value needs an `instanceof` call and will de-specialize.

Reply:
I don't think so; I'm referring to the generated code that Truffle produces when you have an abstract class with specializations, which is a different concern than partial evaluation?